drivers/flash/flash_stm32l4x.c

```diff
@@ -244,6 +245,159 @@ int flash_stm32_write_range(const struct device *dev, unsigned int offset,
 	return rc;
 }
 
+static __unused int write_optb(const struct device *dev, uint32_t mask,
+			       uint32_t value)
+{
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+	int rc;
+
+	if (regs->CR & FLASH_CR_OPTLOCK) {
+		return -EIO;
+	}
+
+	if ((regs->OPTR & mask) == value) {
+		return 0;
+	}
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	regs->OPTR = (regs->OPTR & ~mask) | value;
+	regs->CR |= FLASH_CR_OPTSTRT;
+
+	/* Make sure previous write is completed. */
+	barrier_dsync_fence_full();
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	return 0;
+}
+
+#if defined(CONFIG_FLASH_STM32_WRITE_PROTECT)
+
+/*
+ * STM32L4 allows for 2 write protected memory areas, c.f. FLASH_WEP1AR, FLASH_WRP1BR
+ * which are defined by their start and end pages, i.e. there are two consecutive ranges
+ * the current API passes in two 32 bit mask struct flash_stm32_ex_op_sector_wp_in,
+ * which are barely sufficient to encode the 256 2k pages for the L451 (512kb Flash) part
+ * (embedd all 4 start / end pages number for the two protected areas, each using 8 bits)
+ * this would then be different to STM32F4 meaning of those bits, where they encode single
+ * pages. This could be done here as well, the consecutivity check must be implemented and
+ * thins may be a bit complex regarding the assignement of ranges to FLASH_WEP1AR,
+ * FLASH_WRP1BR
+ */
+
+#endif /* CONFIG_FLASH_STM32_WRITE_PROTECT */
+
+#if defined(CONFIG_FLASH_STM32_READOUT_PROTECTION)
+int flash_stm32_update_rdp(const struct device *dev, bool enable,
+			   bool permanent)
+{
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+	uint8_t current_level, target_level;
+
+	current_level =
+		(regs->OPTR & FLASH_OPTR_RDP_Msk) >> FLASH_OPTR_RDP_Pos;
+	target_level = current_level;
+
+	/*
+	 * 0xAA = RDP level 0 (no protection)
+	 * 0xCC = RDP level 2 (permanent protection)
+	 * others = RDP level 1 (protection active)
+	 */
+	switch (current_level) {
+	case FLASH_STM32_RDP2:
+		if (!enable || !permanent) {
+			__ASSERT(false, "RDP level 2 is permanent and can't be "
+					"changed!");
+			return -ENOTSUP;
```

Comment:
I wonder whether pairing _ASSERT and `return -ENOTSUP` makes sense.
Shouldn't here be rather `ERR_LOG` ?
Assuming that you have __ASSERT enabled for tests only, then why not just instead turn it off and fail the test if `-ENOTSUP` gets returned? You just want to catch problem and never get into it again.
In case when you have __ASSERT enabled in field, by accident, this halts your board.
On the other hand, if you have LOG_ERR here, in case when this fails in field you will have error log but the app may keep executing, so you can pull the logs and figure where the problem is.

Reply:
that code is in sync with drivers/flash/flash_stm32f4x.c module some tricky to spot differences in register names OPTR vs OPTCR. 
If following your suggesting (I see your point here) I would unify those functions and place them into flash_stm32_ex_op_common.c. Would that make sense?
OTOH we could merge as-is and address this in the future.

Reply:
> I would unify those functions and place them into flash_stm32_ex_op_common.c

`flash_stm32_ex_op_common.c` was created for keeping extended operations only. I don't think it's a good place to put here functions that touches the hardware.

Reply:
Regarding `__ASSERT`. It doesn't use flash space when it's not enabled. Flash settings are not changed very often. One can enable asserts to check if there is an error.

Reply:
I force pushed an update addressing this, I also think that LOG_ERR is better suited to __ASSERT